File: impls/erlang/src/step4_if_fn_do.erl
-module(step4_if_fn_do).

-include("types.hrl").

-export([main/1]).

main(_) ->
    Env = env:new(core:ns()),
    rep("(def! not (fn* (a) (if a false true)))", Env),
    repl(Env).

rep(Code, Env) ->
    {success, Ast} = read(Code),
    print(eval(Ast, Env)).

repl(Env) ->
    case io:get_line(standard_io, "user> ") of
        eof ->
            io:format("~n"),
            ok;
        {error, ErrorDescription} ->
            io:format("Error reading input: ~p~n", [ErrorDescription]),
            exit(ioerr);
        Input ->
            case read(Input) of
                {success, Ast} ->
                    try eval(Ast, Env)
                             of Res -> io:format("~s~n", [print(Res)]),
                                       repl(Env)
                    catch
                        _:{malerr, Err} -> io:format("Exception: ~s~n", [Err]),
                                         repl(Env);
                        E:Reason:Stack -> io:format("Mal bug: ~p~n~p~n~p~n", [E, Reason, Stack]),
                                        repl(Env)
                    end;
                {error, Err} ->
                    io:format("Reader error: ~s~n", [Err]),
                    repl(Env)
            end
        end.

read(Code) ->
    reader:read_str(Code).

print(Ast) ->
    printer:pr_str(Ast).

eval(Ast, Env) when not is_record(Ast, mal_list) -> eval_ast(Ast, Env);
eval(#mal_list{elems=[]}, _) -> #mal_list{elems=[]};
eval(AstRec, Env) ->
    AstList = AstRec#mal_list.elems,
    [H|Tail] = AstList,
%    io:format("ast is: ~p~n", [printer:pr_str(AstRec)]),
%    io:format("env is: ~p~n", [Env]),
    case H of
        #mal_sym{val="def!"} ->
            [K, E] = Tail,
            V = eval(E, Env),
            env:set(K, V, Env),
            V;
        #mal_sym{val="let*"} ->
            [Binds, Body] = Tail,
            eval(Body, let_binds(Binds, env:new(Env)));
        #mal_sym{val="do"} ->
            Do = fun (Expr, _) -> eval(Expr, Env) end,
            lists:foldl(Do, mal_nil, Tail);
        #mal_sym{val="if"} ->
            [Cond, Body|Else] = Tail,
            case cond_to_res(eval(Cond, Env)) of
                true  -> eval(Body, Env);
                false -> if length(Else) > 0 -> eval(hd(Else), Env);
                            true -> mal_nil end
            end;
        #mal_sym{val="fn*"} ->
            [Params, Body] = Tail,
            #mal_fn{ast=Body, params=Params, env=Env};
        _ ->
            #mal_list{elems=[F|Args]} = eval_ast(AstRec, Env),
            case F of
                #mal_fn{ast=Body, params=Params, env=CEnv} ->
                    FnEnv = fn_binds(Params, Args, env:new(CEnv)),
                    eval(Body, FnEnv);
                _ ->
                    F(Args)
            end
    end.

let_binds([], Env) -> Env;
let_binds(#mal_list{elems=Binds}, Env) -> let_binds(Binds, Env);
let_binds(#mal_vec{elems=Binds}, Env) -> let_binds(Binds, Env);
let_binds([Sym, Expr|Tail], Env) ->
    env:set(Sym, eval(Expr, Env), Env),
    let_binds(Tail, Env).

fn_binds([], _, Env) -> Env;
fn_binds(#mal_list{elems=Params}, Args, Env) -> fn_binds(Params, Args, Env);
fn_binds(#mal_vec{elems=Params}, Args, Env) -> fn_binds(Params, Args, Env);
fn_binds([#mal_sym{val="&"}, Rest], Args, Env) ->
    env:set(Rest, #mal_list{elems=Args}, Env),
    Env;
fn_binds([Param|Rest], [Arg|Args], Env) ->
    env:set(Param, Arg, Env),
    fn_binds(Rest, Args, Env).

cond_to_res(V) when V =:= mal_false -> false;
cond_to_res(V) when V =:= mal_nil -> false;
cond_to_res(_) -> true.

eval_ast(#mal_list{elems=Elems}, Env) ->
    Acc = fun(A, L) -> L ++ [eval(A, Env)] end,
    #mal_list{elems=lists:foldl(Acc, [], Elems)};
eval_ast(#mal_vec{elems=Elems}, Env) ->
    Acc = fun(A, L) -> L ++ [eval(A, Env)] end,
    #mal_vec{elems=lists:foldl(Acc, [], Elems)};
eval_ast(#mal_map{pairs=Pairs}, Env) ->
    Acc = fun(K, V, M) -> maps:put(K, eval(V, Env), M) end,
    #mal_map{pairs=maps:fold(Acc, #{}, Pairs)};
eval_ast(Ast, Env) when is_record(Ast, mal_sym) ->
    env:get(Ast, Env);
eval_ast(Ast, _) -> Ast.

File: impls/ts/step4_if_fn_do.ts
import { MalList, MalMap, MalType, mal, MalCoreFunc, MalSymbol, MalSeq } from './types.ts'
import { read_str } from './reader.ts'
import { pr_str } from './printer.ts'
import Env from './env.ts';
import core from './core.ts';

function eval_ast(ast: MalType, env: Env): MalType {
    switch(ast.type) {
        case 'symbol': {
            return env.get(ast)
        }
        case 'list':
        case 'vector': {
            const valGen = mal[ast.type]
            const seed = valGen([])
                return ast.values.reduce(
                    (a: typeof seed, b: MalType) => valGen(a.values.concat([EVAL(b, env)])),
                    seed
                )
            }
        case 'map': {
            return Array.from(ast.values.entries()).reduce(
                (acc: MalMap, [k, v]) => mal.map(acc.values.set(k, EVAL(v, env))),
                mal.map(new Map())
            )
        }
        default:
            return ast
    }
}

function READ(s: string): MalType {
    return read_str(s)
}
function EVAL(v: MalType, env: Env): MalType {
    if(v.type != 'list') {
        return eval_ast(v, env)
    }
    else if(v.values.length === 0) {
        return v
    }
    else {
        const head = v.values[0]
        const tail = v.values.slice(1)

        if(head.type === 'symbol') {
            switch(head.value) {
                case 'def!': {
                    const v = EVAL(tail[1], env)
                    return env.set((tail[0] as MalSymbol), v)
                }
                case 'let*': {
                    return EVAL(tail[1], makeEnv((tail[0] as MalList), env))
                }
                case 'do': {
                    const result = eval_ast(mal.list(tail), env) as MalList
                    return result.values.pop()!
                }
                case 'if': {
                    const guard = EVAL(tail[0], env)
                    if(isTrue(guard)) {
                        return EVAL(tail[1], env)
                    }
                    else if (tail.length > 2) {
                        return EVAL(tail[2], env)
                    }
                    else {
                        return mal.nil()
                    }
                }
                case 'fn*': {
                    const binds = tail[0] as MalSeq
                    const body  = tail[1]
                    return mal.function((...vals: MalType[]) => {
                        return EVAL(body, new Env(env, binds, mal.list(vals)))
                    })
                }
            }
        }

        const callSite = eval_ast(v, env) as MalList
        const f        = callSite.values[0] as MalCoreFunc
        const args     = callSite.values.slice(1)

        return f.value.apply(null, args)
    }
}
function PRINT(v: MalType): string {
    return pr_str(v)
}
function rep(s: string): string {
    return PRINT(EVAL(READ(s), env))
}

function makeEnv(pairs: MalList, outerEnv: Env): Env {
    const newEnv = new Env(outerEnv)
    for(let idx = 0; idx < pairs.values.length; idx += 2) {
        const k = pairs.values[idx] as MalSymbol
        const v = pairs.values[idx + 1]
        newEnv.set(k, EVAL(v, newEnv))
    }
    return newEnv
}

function isTrue(v: MalType): boolean {
    return !(v.type == 'bool' && !v.value || v.type == 'nil')
}

const env = new Env()
Object.entries(core).forEach(([s, f]) => {
    env.set(mal.symbol(s), f)
})

rep("(def! not (fn* (a) (if a false true)))")

let line = prompt('user>')
while (line !== null) {
    try {
        console.log(rep(line));
    }
    catch(e) {
        console.log('Error: ', e)
    }
    line = prompt('user>')
}
